services/src/main/java/org/keycloak/authorization/admin/representation/PolicyEvaluationResponseBuilder.java

```diff
@@ -55,6 +56,12 @@ public static PolicyEvaluationResponse build(List<Result> results, ResourceServe
         authorizationData.setPermissions(Permissions.permits(results, null, authorization, resourceServer));
         accessToken.setAuthorization(authorizationData);
 
+        ClientModel clientModel = authorization.getRealm().getClientById(resourceServer.getId());
+
+        if (!accessToken.hasAudience(clientModel.getClientId())) {
+            accessToken.audience(clientModel.getClientId());
+        }
```

Comment:
Is this logic of "overwrite all existing audiences with the client id if not contained but keep all if contained" really intended? I would have expected to either just add the audience if not contained or basically always use the client id as audience only.

Reply:
Actually, it's a security issue to add the client as audience in the accessToken. The client is rightful audience for the IdToken, but accessToken should be opaque to the client.

Reply:
In this case the client is actually the resource server (see ClientModel clientModel = authorization.getRealm().getClientById(resourceServer.getId());)

Reply:
Keycloak does not support multiple audiences in an access token, so the the solution should be fine.

And yes, the {{clientModel}} is always the resource server.

Reply:
The implementation suggests otherwise as aud is a string array and audience() is variadic, see JsonWebToken.java:
@JsonProperty("aud")
@JsonSerialize(using = StringOrArraySerializer.class)
@JsonDeserialize(using = StringOrArrayDeserializer.class)
protected String[] audience;

public JsonWebToken audience(String... audience) {
        this.audience = audience;
        return this;
    }

public boolean hasAudience(String audience) {
        if (this.audience == null) return false;
        for (String a : this.audience) {
            if (a.equals(audience)) {
                return true;
            }
        }
        return false;
    }
Even if it is always a one-element array currently, I am concerned it will come up as a bug if multiple audiences get supported.

Reply:
Yeah, you are right. The thing is that audience is something we are failing to do correctly. From token issuance to validation (e.g.: by built-in adapters) we are not dealing with it correctly.

That would be a quite bug change to existing code. There are some issues at this respect.

My point is that for now, the changes in this PR should solve our initial problem. Where in the future we might want to revisit this part (and others) to properly handle audience.

Reply:
I am fine with that. Best place to support adding or removing audiences is probably in the token class anyway. You might want to add a TODO somewhere to revisit this once multiple audiences are supported more properly.

Reply:
OK, we are tracking this on JIRA. Thanks for all feedback.